### PR TITLE
Windows compilation: enable compiling expanded list of extensions in envoy-static

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Description:
- replace WINDOWS_EXTENSIONS with very short WINDOWS_SKIP_TARGETS list
    - extensions _SKIP_TARGET lists converted from dict to list
    - See https://github.com/envoyproxy/envoy/pull/10365 for attempted
      workaround development, we were unsuccessful at a short-term
      fix to integrate WINDOWS_SKIP_TARGETS / PPC_SKIP_TARGETS
- fixes kafka compilation (MSVC parses the literal of INT32_MIN incorrectly)
- fixes luajit and moonjit compilation (only one jit is elected)
- merge windows .bazelrc into workspace .bazelrc
- install MSYS2 into the CI worker to ensure the expected shell environment for bazel and build tools are 
  correctly on the PATH
- Building/running tests in CI is not yet enabled, we are investigating options related to https://github.com/envoyproxy/envoy/issues/10619

Risk Level: Low (Windows compilation only, small source changes)
Testing: N/A, just enabling Windows compilation
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by sunjayBhatia** [comment]
Running into Windows Azure CI environment issues, seems like the `link` being found on the `PATH` is not the correct `link.exe` from MSVC, we are seeing wht we can do about this scenario

**Comment by sunjayBhatia** [comment]
It looks like git bash is the source of our problems, we're going to install msys2 in our CI environment in the setup phase and rely on that instead of fighting git bash

Hopefully, according to https://github.com/actions/virtual-environments/pull/632 msys2 will be installed in the base image and we can rely on that installation in the future

**Comment by jmarantz** [comment]
hi -- can you ping when this is ready for review?

**Comment by wrowe** [comment]
CI is now unblocked, our issues are resolved.

Pinging @alyssawilk @htuch @jmarantz @lizan @mattklein123  for code review

**Comment by sunjayBhatia** [comment]
not sure why circleci is failing

seeing:
```
ci/build_setup.sh: line 74:    81 Hangup                  "$(dirname "$0")"/../bazel/setup_clang.sh "${LLVM_ROOT}"

Too long with no output (exceeded 10m0s): context deadline exceeded
```

and:
```
cp: cannot stat '/source/*.bazelrc': No such file or directory
```

/retest

**Comment by lizan** [comment]
Let's ignore clang-tidy for now, it picks up master failures because of the number of files changed.

File: source/extensions/filters/http/adaptive_concurrency/BUILD [hunk 1]
```diff
@@ -20,7 +20,7 @@ envoy_cc_library(
     deps = [
         "//include/envoy/http:filter_interface",
         "//source/extensions/filters/http:well_known_names",
-        "//source/extensions/filters/http/adaptive_concurrency/concurrency_controller:concurrency_controller_lib",
+        "//source/extensions/filters/http/adaptive_concurrency/controller:controller_lib",
```
Comment (lizan) [inline source/extensions/filters/http/adaptive_concurrency/BUILD]: ? This refactor part should be in a separate PR.

File: ci/windows_ci_steps.sh [hunk 2]
```diff
@@ -20,14 +20,22 @@ df -h
...
 
-bazel ${BAZEL_STARTUP_OPTIONS} build ${BAZEL_BUILD_OPTIONS} //bazel/... --build_tag_filters=-skip_on_windows
+# With all envoy-static and //test/ tree building, no need to test compile externals
+# bazel ${BAZEL_STARTUP_OPTIONS} build ${BAZEL_BUILD_OPTIONS} //bazel/... --build_tag_filters=-skip_on_windows
 
 bazel ${BAZEL_STARTUP_OPTIONS} build ${BAZEL_BUILD_OPTIONS} //source/exe:envoy-static --build_tag_filters=-skip_on_windows
 
-# bazel ${BAZEL_STARTUP_OPTIONS} test ${BAZEL_BUILD_OPTIONS} //test/... --test_tag_filters=-skip_on_windows --build_tests_only --test_summary=terse --test_output=errors
+# Test compilation of known MSVC-compatible test sources
+bazel ${BAZEL_STARTUP_OPTIONS} build ${BAZEL_BUILD_OPTIONS} //test/... --test_tag_filters=-skip_on_windows --build_tests_only
```
Comment (lizan) [inline ci/windows_ci_steps.sh]: Why just build but not run?

File: source/extensions/filters/http/adaptive_concurrency/BUILD (hunk 1 above)
Comment (sunjayBhatia) [reply source/extensions/filters/http/adaptive_concurrency/BUILD]: This is required for compilation, the path length is an issue that causes headers not to be found by `cl`, this change is just an effort to reduce the length of the path to these files

File: ci/windows_ci_steps.sh (hunk 2 above)
Comment (sunjayBhatia) [reply ci/windows_ci_steps.sh]: This change was just an effort to ensure compilation of test sources is checked in CI. Not all the tests pass yet, we have not yet started delving into the test failures/figuring out what is flaky/fixing them fully. We didn't want to run all tests, have them fail intermittently, and break other PRs/pushes. If desired, we can tag all the tests that fail with `fails_on_windows` (no guarantees we find all the flaky ones).

File: source/extensions/filters/http/adaptive_concurrency/BUILD (hunk 1 above)
Comment (sunjayBhatia) [reply source/extensions/filters/http/adaptive_concurrency/BUILD]: See: https://github.com/envoyproxy/envoy/pull/10542/commits/e7f9c0599ad86cc524d03b0e6d9a70236a35d1ef

Comment (wrowe) [reply source/extensions/filters/http/adaptive_concurrency/BUILD]: It's already in a separate PR, simple for us to pull it out, Thanks!

File: ci/windows_ci_steps.sh (hunk 2 above)
Comment (wrowe) [reply ci/windows_ci_steps.sh]: Right now, about 73 tests fail. After the next follow-up PR, we expect that about 5 corrections and abstractions will resolve 80% of those failures. Trying to keep the scope of each PR review reasonably manageable.

We've been increasing the stress on the wider community slowly. Thus far, we are simply trying to ensure new compilation breakage isn't introduced due to clang/gcc assumptions that MSVC doesn't interpret in the same way, and have been assisting PR contributors with guidance in resolving these compilation discrepancies for envoy-static. We'll continue to do so with the newly introduced //test/... compilation tests.

The resulting behavior differences of the Windows architecture begin to be addressed in next major PR, with a basket of remaining broken tests marked fails_on_windows on a short-term basis, to keep focus on not introducing regressions. The remaining 20% of failures will each demand more detailed fixes (likely 1 PR per failure) and closer scrutiny.

The end goal is 100% success and relies on additional volunteers (outside of our company) who are joining the effort already.

File: source/extensions/filters/http/adaptive_concurrency/BUILD (hunk 1 above)
Comment (wrowe) [reply source/extensions/filters/http/adaptive_concurrency/BUILD]: Although it wasn't quite that simple, https://github.com/envoyproxy/envoy/pull/10560 factors this out of the patch. Please expedite that review and it will all be picked up when we merge this patch back to master. As Sunjay points out, it's required in order for *this* PR to pass compilation CI.

**Comment by wrowe** [comment]
It appears that 2 CPU's and insufficient memory aren't going to allow us to compile, never mind execute the tests. @lizan do you have any thoughts on the envoy-windows failure? Looks straightforward to remedy with correct VM configuration.

**Comment by sunjayBhatia** [comment]
Yeah, we may have to bump the resources given to the windows worker or possible tweak how much memory bazel can take up, at the moment we have the same setup as the rest of the CI environments which could be relevant: https://github.com/envoyproxy/envoy/blob/e9248e242d18e8754d2c07272a8fbae5a8a6b47a/windows/.bazelrc#L7

**Comment by lizan** [comment]
You don't need that large heap for Bazel on Windows because of the parallelism is small (2) compared to Linux (60 - 75). 512m should be more than enough.

**Comment by wrowe** [comment]
> You don't need that large heap for Bazel on Windows because of the parallelism is small (2) compared to Linux (60 - 75). 512m should be more than enough.

Looking at build speed, I'm expecting we will need 4 cores (right now, we build locally with 8 on 8GB.) Reduced as suggested.

File: source/extensions/filters/http/adaptive_concurrency/BUILD (hunk 1 above)
Comment (wrowe) [reply source/extensions/filters/http/adaptive_concurrency/BUILD]: And this is live on master. Thanks!

**Comment by wrowe** [comment]
Looks like https://circleci.com/gh/envoyproxy/envoy coverage tests are almost universally failing today? bazel clang_tidy is a known issue as discussed on slack. However...

```
2020-03-30T23:40:35.7792593Z [194 / 208] Compiling test/tools/router_check/coverage.cc; 49s local ... (2 actions running)
2020-03-30T23:41:28.8706636Z ERROR: D:/a/1/s/test/tools/router_check/BUILD:18:1: C++ compilation of rule '//test/tools/router_check:router_check_main_lib' failed (Exit 2): cl.exe failed: error executing command 
2020-03-30T23:41:28.9050624Z external/com_google_absl\absl/container/internal/hash_policy_traits.h(77): fatal error C1060: compiler is out of heap space
```

It appears we still need to provision 2x memory, even with a reduced 512m java heap, and given that it dies 2 hours in, doubling CPU (4 work threads) at the same time is probably worthwhile.

**Comment by sunjayBhatia** [comment]
Curiously, on a local dev VM with 2 CPUs and 7GB RAM, we're able to get the whole build to complete (though it is *slow*)

**Comment by sunjayBhatia** [comment]
@lizan we've disabled compiling tests in CI for now as it is not consistently successful and takes forever, we're looking into RBE for Windows (here: https://github.com/envoyproxy/envoy/issues/10619) and a last resort working with self managed AZP workers so we think we are ready for this review to continue if you can!

**Comment by wrowe** [comment]
After correcting for EXTENSIONS and restoring the coverage regression, we are left with some unexpected windows test compilation issues. We'll have these addressed shortly.

/wait

**Comment by wrowe** [comment]
Please wait on review (won't tell the bot until after we inspect the compilation results) of this pass of new fixes. Expecting to still push one more set of patches for final review.

**Comment by sunjayBhatia** [comment]
Also pulled some changes out into https://github.com/envoyproxy/envoy/pull/10780 for discussion

**Comment by wrowe** [comment]
@jmarantz and fellow reviewers, this is ready for review, we found a minor mistake in comments which we will be committing shortly. The tests have all passed at one point (at this moment the trouble exists with the CI coverage environment). The net effect of all parts of this patch on linux builds should be null.

Worried that the patch might appear to be intimidating to review, would it be helpful if we pulled all of the .cc/.h source code changes from source/ and test/ and present those separately as a single PR?

**Comment by wrowe** [comment]
Before removing test/* changes from this PR, all tests passing, except for coverage (known to be unstable.)

**Comment by sunjayBhatia** [comment]
Update the title and description with the more limited scope for this PR, let us know if we should pare this set of changes down more for a more manageable review.

**Comment by mattklein123** [comment]
Please merge master. Coverage should be fixed now (needs master merge anyway).

/wait

**Comment by wrowe** [comment]
/retest

for macOS

**Comment by wrowe** [comment]
It doesn't seem possible to get through the macOS build step due to the current CI situation. Otherwise this appears to be looking good for acceptance.

**Comment by mattklein123** [comment]
Thanks this LGTM. Can you please merge master? OSX is reliably passing now on master. Thank you!

/wait

**Comment by wrowe** [comment]
/retest
... to show coverage publish completion
